functional-tests/appwrapper_reconciler_test.go

```diff
@@ -0,0 +1,250 @@
+		test.Expect(err).ToNot(HaveOccurred())
+		return machineset
+	}
+}
+
+func MachineSetReplicas(machineSet *machinev1beta1.MachineSet) *int32 {
+	return machineSet.Spec.Replicas
+}
+
+func TestReconciler(t *testing.T) {
+	// initiate setting up the EnvTest environment
+	startEnvTest(t)
+
+	test := WithConfig(t, cfg)
+
+	// initialize a variable with test client
+	client := test.Client()
+
+	//read machineset yaml from file `test_instascale_machineset.yml`
+	b, err := os.ReadFile("test_instascale_machineset.yml")
+	test.Expect(err).ToNot(HaveOccurred())
+
+	//deserialize kubernetes object
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+	ms, _, err := decode(b, nil, nil) //decode machineset content of YAML file into kubernetes object
+	test.Expect(err).ToNot(HaveOccurred())
+	msa := ms.(*machinev1beta1.MachineSet) //asserts that decoded object is of type `*machinev1beta1.MachineSet`
+
+	//create machineset in default namespace
+	ms, err = client.Machine().MachineV1beta1().MachineSets("default").Create(test.Ctx(), msa, metav1.CreateOptions{})
+	test.Expect(err).ToNot(HaveOccurred())
+
+	//assert that the replicas in Machineset specification is 0, before appwrapper creation
+	test.Eventually(MachineSet(test, "default", "test-instascale")).Should(WithTransform(MachineSetReplicas, gstruct.PointTo(Equal(int32(0)))))
+
+	//create new test namespace
+	namespace := test.NewTestNamespace()
+
+	// initializes an appwrapper in the created namespace
+	aw := instascaleAppwrapper(namespace.Name)
+
+	// create appwrapper resource using mcadClient
+	aw, err = client.MCAD().WorkloadV1beta1().AppWrappers(namespace.Name).Create(test.Ctx(), aw, metav1.CreateOptions{})
+	test.Expect(err).ToNot(HaveOccurred())
+	time.Sleep(10 * time.Second)
+
+	//update appwrapper status to avoid appwrapper dispatch in case of Insufficient resources
+	aw.Status = mcadv1beta1.AppWrapperStatus{
+		Pending: 1,
+		State:   mcadv1beta1.AppWrapperStateEnqueued,
+		Conditions: []mcadv1beta1.AppWrapperCondition{
+			{
+				Type:    mcadv1beta1.AppWrapperCondBackoff,
+				Status:  apiv1.ConditionTrue,
+				Reason:  "AppWrapperNotRunnable",
+				Message: "Insufficient resources to dispatch AppWrapper.",
+			},
+		},
+	}
+	_, err = client.MCAD().WorkloadV1beta1().AppWrappers(namespace.Name).UpdateStatus(test.Ctx(), aw, metav1.UpdateOptions{})
+	test.Expect(err).ToNot(HaveOccurred())
+	time.Sleep(10 * time.Second)
+
+	// assert for machine replicas belonging to the machine set after appwrapper creation- there should be 1
+	test.Eventually(MachineSet(test, "default", "test-instascale")).Should(WithTransform(MachineSetReplicas, gstruct.PointTo(Equal(int32(1)))))
+
+	// delete appwrapper
+	err = client.MCAD().WorkloadV1beta1().AppWrappers(namespace.Name).Delete(test.Ctx(), aw.Name, metav1.DeleteOptions{})
+	test.Expect(err).ToNot(HaveOccurred())
+	time.Sleep(10 * time.Second)
```

Comment:
Can you remove this line? It shouldn't be needed.

Reply:
In case of verifying if machineSet replicas scale down do 0 once apwrapper is deleted, for this verification we added this step right?

Reply:
In case machineSet replicas can be scaled down to 0 once apwrapper is deleted, to verify this scenario we added this step, right? After this step there I have added one more check to see if number of replicas scale down to 0 or not ```  // assert for machine belonging to the machine set after appwrapper deletion- there should be none
	test.Eventually(MachineSet(test, "default", "test-instascale")).Should(WithTransform(MachineSetReplicas, gstruct.PointTo(Equal(int32(0))))) 
```